"tls: internal error" when using GetConfigForClient

First off, just wanted to thank you again for working on this lib. I've been fiddling around with integrating it into Caddy 2 this weekend, with moderate success!

Unlike in Caddy 1 with gQUIC, I will be treating IETF HTTP/3 as a first-class feature in Caddy 2 -- not experimental (well, once it gets standardized). But the HTTP/3 support will get a lot more love and attention than the gQUIC support did in Caddy 1. We intend to use the master branch -- not gQUIC -- for Caddy 2.

Anyway, I seem to have a problem that _appears_ to be a bit of a showstopper, unless you know a good workaround or have a fix in mind.

I've been using the client example in this repo to test the HTTP/3 support.

It seems that when I provide a TLS config to the HTTP/3 _server_ with a static certificate and key file, it works fine:


```
$ GO111MODULE=on go run main.go -insecure https://localhost:2443
GET https://localhost:2443
client Starting new connection to localhost ([::]:59085 -> 127.0.0.1:2443), source connection ID (empty), destination connection ID 0x99986f151ffc92d4019ba5, version QUIC WG draft-22
Got response for https://localhost:2443: &http.Response{Status:"200 OK", StatusCode:200, Proto:"HTTP/3", ProtoMajor:3, ProtoMinor:0, Header:http.Header{}, Body:(*http3.body)(0xc0001adf40), ContentLength:0, TransferEncoding:[]string(nil), Close:false, Uncompressed:false, Trailer:http.Header(nil), Request:(*http.Request)(nil), TLS:(*tls.ConnectionState)(nil)}
Request Body:
Hello, HTTP/3. Caddy at your service.
client Closing session.
client Connection (empty) closed.
```

Yay!

But certificates expire and stuff.

So, Caddy dynamically handles certificates using the tls.Config.GetConfigForClient field. In other words, we don't populate the Certificates field at all. But all other things being the same, not filling the Certificates field leads to this error:

```
GO111MODULE=on go run main.go -insecure https://localhost:2443
GET https://localhost:2443
client Starting new connection to localhost ([::]:59084 -> 127.0.0.1:2443), source connection ID (empty), destination connection ID 0xa1cb57c374c0e25eaf34d8bd951ae6be, version QUIC WG draft-22
client Peer closed session with error: CRYPTO_ERROR: tls: internal error
client Connection (empty) closed.
panic: Get https://localhost:2443: CRYPTO_ERROR: tls: internal error

goroutine 18 [running]:
main.main.func1(0xc0000a2450, 0x153f960, 0xc0000c4ab0, 0xc0000ce591, 0xc0000b4370, 0x7ffeefbffa1b, 0x16)
	/Users/matt/Dev/src/github.com/lucas-clemente/quic-go/example/client/main.go:50 +0x435
created by main.main
	/Users/matt/Dev/src/github.com/lucas-clemente/quic-go/example/client/main.go:47 +0x43b
exit status 2
```

Any chance it can work with GetConfigForClient? Pretty please? :smile:

PS. I'd also be happy to make Caddy 2 a proper test bed for production deployments of HTTP/3, once we get this sorted out. Integrating it seems easy enough otherwise! (We've had some issues with reusing listeners in the past, but maybe that'll be resolved with https://godoc.org/github.com/lucas-clemente/quic-go/http3#Server.CloseGracefully.)

--- a/http3/client.go
+++ b/http3/client.go
@@ -57,10 +57,11 @@ func newClient(
 ) *client {
 	if tlsConf == nil {
 		tlsConf = &tls.Config{}
+	} else {
+		tlsConf = tlsConf.Clone()
 	}
-	if !strSliceContains(tlsConf.NextProtos, nextProtoH3) {
-		tlsConf.NextProtos = append(tlsConf.NextProtos, nextProtoH3)
-	}
+	// Replace existing ALPNs by H3
+	tlsConf.NextProtos = []string{nextProtoH3}
 	if quicConfig == nil {
 		quicConfig = defaultQuicConfig
 	}

--- a/http3/client_test.go
+++ b/http3/client_test.go
@@ -85,13 +85,15 @@ var _ = Describe("Client", func() {
 		) (quic.Session, error) {
 			Expect(hostname).To(Equal("localhost:1337"))
 			Expect(tlsConfP.ServerName).To(Equal(tlsConf.ServerName))
-			Expect(tlsConfP.NextProtos).To(Equal([]string{"proto foo", "proto bar", nextProtoH3}))
+			Expect(tlsConfP.NextProtos).To(Equal([]string{nextProtoH3}))
 			Expect(quicConfP.IdleTimeout).To(Equal(quicConf.IdleTimeout))
 			dialAddrCalled = true
 			return nil, errors.New("test done")
 		}
 		client.RoundTrip(req)
 		Expect(dialAddrCalled).To(BeTrue())
+		// make sure the original tls.Config was not modified
+		Expect(tlsConf.NextProtos).To(Equal([]string{"proto foo", "proto bar"}))
 	})
 
 	It("uses the custom dialer, if provided", func() {
@@ -102,7 +104,7 @@ var _ = Describe("Client", func() {
 		dialer := func(network, address string, tlsConfP *tls.Config, quicConfP *quic.Config) (quic.Session, error) {
 			Expect(network).To(Equal("udp"))
 			Expect(address).To(Equal("localhost:1337"))
-			Expect(tlsConfP).To(Equal(tlsConf))
+			Expect(tlsConfP.ServerName).To(Equal("foo.bar"))
 			Expect(quicConfP.IdleTimeout).To(Equal(quicConf.IdleTimeout))
 			dialerCalled = true
 			return nil, testErr

--- a/http3/server.go
+++ b/http3/server.go
@@ -99,6 +99,17 @@ func (s *Server) serveImpl(tlsConf *tls.Config, conn net.PacketConn) error {
 	}
 	// Replace existing ALPNs by H3
 	tlsConf.NextProtos = []string{nextProtoH3}
+	if tlsConf.GetConfigForClient != nil {
+		getConfigForClient := tlsConf.GetConfigForClient
+		tlsConf.GetConfigForClient = func(ch *tls.ClientHelloInfo) (*tls.Config, error) {
+			conf, err := getConfigForClient(ch)
+			if err != nil || conf == nil {
+				return conf, err
+			}
+			conf.NextProtos = []string{nextProtoH3}
+			return conf, nil
+		}
+	}
 
 	var ln quic.Listener
 	var err error

--- a/http3/server.go
+++ b/http3/server.go
@@ -77,7 +77,7 @@ func (s *Server) Serve(conn net.PacketConn) error {
 	return s.serveImpl(s.TLSConfig, conn)
 }
 
-func (s *Server) serveImpl(tlsConfig *tls.Config, conn net.PacketConn) error {
+func (s *Server) serveImpl(tlsConf *tls.Config, conn net.PacketConn) error {
 	if s.Server == nil {
 		return errors.New("use of http3.Server without http.Server")
 	}
@@ -92,20 +92,20 @@ func (s *Server) serveImpl(tlsConfig *tls.Config, conn net.PacketConn) error {
 		return errors.New("ListenAndServe may only be called once")
 	}
 
-	if tlsConfig == nil {
-		tlsConfig = &tls.Config{}
+	if tlsConf == nil {
+		tlsConf = &tls.Config{}
+	} else {
+		tlsConf = tlsConf.Clone()
 	}
-
-	if !strSliceContains(tlsConfig.NextProtos, nextProtoH3) {
-		tlsConfig.NextProtos = append(tlsConfig.NextProtos, nextProtoH3)
-	}
+	// Replace existing ALPNs by H3
+	tlsConf.NextProtos = []string{nextProtoH3}
 
 	var ln quic.Listener
 	var err error
 	if conn == nil {
-		ln, err = quicListenAddr(s.Addr, tlsConfig, s.QuicConfig)
+		ln, err = quicListenAddr(s.Addr, tlsConf, s.QuicConfig)
 	} else {
-		ln, err = quicListen(conn, tlsConfig, s.QuicConfig)
+		ln, err = quicListen(conn, tlsConf, s.QuicConfig)
 	}
 	if err != nil {
 		s.listenerMutex.Unlock()
@@ -384,13 +384,4 @@ func ListenAndServe(addr, certFile, keyFile string, handler http.Handler) error 
 		// Cannot close the HTTP server or wait for requests to complete properly :/
 		return err
 	}
-}
-
-func strSliceContains(ss []string, s string) bool {
-	for _, v := range ss {
-		if v == s {
-			return true
-		}
-	}
-	return false
 }

--- a/http3/server_test.go
+++ b/http3/server_test.go
@@ -406,6 +406,30 @@ var _ = Describe("Server", func() {
 			Expect(s.ListenAndServe()).To(HaveOccurred())
 			Expect(receivedConf.NextProtos).To(Equal([]string{nextProtoH3}))
 		})
+
+		It("sets the ALPN for tls.Configs returned by the tls.GetConfigForClient", func() {
+			tlsConf := &tls.Config{
+				GetConfigForClient: func(ch *tls.ClientHelloInfo) (*tls.Config, error) {
+					return &tls.Config{NextProtos: []string{"foo", "bar"}}, nil
+				},
+			}
+
+			var receivedConf *tls.Config
+			quicListenAddr = func(addr string, conf *tls.Config, _ *quic.Config) (quic.Listener, error) {
+				receivedConf = conf
+				return nil, errors.New("listen err")
+			}
+			s.TLSConfig = tlsConf
+			Expect(s.ListenAndServe()).To(HaveOccurred())
+			// check that the config used by QUIC uses the h3 ALPN
+			conf, err := receivedConf.GetConfigForClient(&tls.ClientHelloInfo{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(conf.NextProtos).To(Equal([]string{nextProtoH3}))
+			// check that the original config was not modified
+			conf, err = tlsConf.GetConfigForClient(&tls.ClientHelloInfo{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(conf.NextProtos).To(Equal([]string{"foo", "bar"}))
+		})
 	})
 
 	Context("ListenAndServeTLS", func() {

--- a/http3/server_test.go
+++ b/http3/server_test.go
@@ -383,15 +383,18 @@ var _ = Describe("Server", func() {
 			Expect(receivedConf).To(Equal(conf))
 		})
 
-		It("adds the ALPN token to the tls.Config", func() {
+		It("replaces the ALPN token to the tls.Config", func() {
+			tlsConf := &tls.Config{NextProtos: []string{"foo", "bar"}}
 			var receivedConf *tls.Config
 			quicListenAddr = func(addr string, tlsConf *tls.Config, _ *quic.Config) (quic.Listener, error) {
 				receivedConf = tlsConf
 				return nil, errors.New("listen err")
 			}
-			s.TLSConfig = &tls.Config{NextProtos: []string{"foo", "bar"}}
+			s.TLSConfig = tlsConf
 			Expect(s.ListenAndServe()).To(HaveOccurred())
-			Expect(receivedConf.NextProtos).To(Equal([]string{"foo", "bar", nextProtoH3}))
+			Expect(receivedConf.NextProtos).To(Equal([]string{nextProtoH3}))
+			// make sure the original tls.Config was not modified
+			Expect(tlsConf.NextProtos).To(Equal([]string{"foo", "bar"}))
 		})
 
 		It("uses the ALPN token if no tls.Config is given", func() {

